crema/qvalues.py

```diff
@@ -132,3 +132,192 @@ def _fdr2qvalue(scores, fdr):
+        A list of p-values sorted from smallest to largest.
+
+    Returns
+    -------
+    pi0 : float
+        Estimated pi_zero.
+    """
+    numLambda = 100
+    maxLambda = 0.5
+    numBoot = 100
+
+    n_pval = pval_list.size
+    lambda_list = []
+    pi0s_list = []
+    for idx in range(0, numLambda):
+        cur_lambda = ((idx + 1) / numLambda) * maxLambda
+
+        # Find the index of the first element in pValList
+        # that is > lambda.
+        start = np.searchsorted(pval_list, cur_lambda)
+        W1 = n_pval - start
+        pi0 = W1 / n_pval / (1.0 - cur_lambda)
+
+        if pi0 > 0.0:
+            lambda_list.append(cur_lambda)
+            pi0s_list.append(pi0)
+
+    assert len(pi0s_list) != 0, (
+        "Error in the input data: "
+        "too good separation between target and decoy PSMs."
+    )
+
+    minPi0 = min(pi0s_list)
+
+    mse_list = np.zeros(len(pi0s_list))
+    max_size = 1000
+    # Examine which lambda level that is most stable under bootstrap
+    for i in range(0, numBoot):
+        # Create an array of bootstrapped p-values, and sort in ascending order.
+        num_draw = min(n_pval, max_size)
+        pBoot_list = np.random.choice(pval_list, size=num_draw, replace=True)
+        pBoot_list.sort()
+
+        for idx in range(0, len(lambda_list)):
+            start = np.searchsorted(pBoot_list, lambda_list[idx])
+            W1 = num_draw - start
+            pi0Boot = W1 / num_draw / (1.0 - lambda_list[idx])
+
+            # Estimated mean-squared error
+            mse_list[idx] += (pi0Boot - minPi0) * (pi0Boot - minPi0)
+
+    # Which index did the iterator go?
+    minIdx = np.argmin(mse_list)
+    pi0 = max(min(pi0s_list[minIdx], 1.0), 0.0)
+    return pi0
+
+
+@nb.njit
+def calculate_mixmax_qval(target_scores, decoy_scores, pi0):
+    """ """
+    num_targets = target_scores.shape[0]
+    num_decoys = decoy_scores.shape[0]
+
+    h_w_le_z = np.zeros(num_decoys + 1)  # histogram for N_{w<=z}
+    h_z_le_z = np.zeros(num_decoys + 1)  # histogram for N_{z<=z}
+
+    for i in range(0, num_decoys):
+        h_w_le_z[i] = np.searchsorted(
+            target_scores, decoy_scores[i], side="right"
+        )
+        h_z_le_z[i] = np.searchsorted(
+            decoy_scores, decoy_scores[i], side="right"
+        )
+
+    h_w_le_z[num_decoys] = num_targets
+    h_z_le_z[num_decoys] = num_decoys
+
+    fdrmod = np.zeros(num_targets)
+    estPx_lt_zj = 0.0
+    E_f1_mod_run_tot = 0.0
+    j = num_decoys - 1
+    k = num_targets - 1
+    n_z_ge_w = 0
+    n_w_ge_w = 0
+    prev_fdr = -1
+
+    for i in range(num_targets - 1, -1, -1):
+        while j >= 0 and decoy_scores[j] >= target_scores[i]:
+            cnt_w = h_w_le_z[j + 1]
+            cnt_z = h_z_le_z[j + 1]
+
+            estPx_lt_zj = (cnt_w - pi0 * cnt_z) / ((1.0 - pi0) * cnt_z)
+            if estPx_lt_zj > 1:
+                estPx_lt_zj == 1.0
+            elif estPx_lt_zj < 0:
+                estPx_lt_zj == 0.0
```

Comment:
I believe this is meant to clamp the value to `[0, 1]`, so `==` (comparison) should be `=` (assignment).

Reply:
Yup, I think are you correct. Thanks for catching that. I have made the change.